### PR TITLE
FxIOS-8422 Add Edit button to Bookmarks folders

### PR DESCRIPTION
**Pull request by electricRGB**

### CODE REVIEW

File: Client/Frontend/Library/LibraryViewController/LibraryViewController.swift [hunk 1]
```diff
@@ -18,11 +18,25 @@ extension LibraryViewController: UIToolbarDelegate {
     }
 }
 
+/// This enum describes the different states the Bookmarks panel,
+/// in the Library Panel, can have. All other Library Panel's do
+/// do not have states associated with them, allowing this one
```
Comment (nbhasin2) [inline Client/Frontend/Library/LibraryViewController/LibraryViewController.swift:23]:
```suggestion
/// in the Library Panel, can have. All other Library Panels do
/// not have states associated with them, allowing this one
```

File: Client/Frontend/Library/LibraryViewController/LibraryViewController.swift [hunk 2]
```diff
@@ -217,22 +231,113 @@ class LibraryViewController: UIViewController {
...
+            bookmarksPanel.enableEditMode()
+
+        case .inFolderEditMode:
+            guard let bookmarksPanel = panel.viewControllers.last as? BookmarksPanel else { return }
+            bookmarkPanelState = .inFolder
+            bookmarksPanel.disableEditMode()
+
+        case .bookmarkEditMode:
+            guard let bookmarkEditView = panel.viewControllers.last as? BookmarkDetailPanel else { return }
+
```
Comment (nbhasin2) [inline Client/Frontend/Library/LibraryViewController/LibraryViewController.swift:329]:
```suggestion
```

nit: extra space

File: Client/Frontend/Library/LibraryViewController/LibraryViewController.swift [hunk 3]
```diff
@@ -217,22 +231,113 @@ class LibraryViewController: UIViewController {
...
+        switch bookmarkPanelState {
+        case .home:
+            self.dismiss(animated: true, completion: nil)
+
+        case .inFolder:
+            guard let bookmarksPanel = panel.viewControllers.last as? BookmarksPanel else { return }
+            bookmarkPanelState = .inFolderEditMode
+            bookmarksPanel.enableEditMode()
+
```
Comment (nbhasin2) [inline Client/Frontend/Library/LibraryViewController/LibraryViewController.swift:321]:
```suggestion
```
nit: extra space

File: Client/Frontend/Library/LibraryViewController/LibraryViewController.swift [hunk 4]
```diff
@@ -217,22 +231,113 @@ class LibraryViewController: UIViewController {
...
-    
+
     @objc func topRightButtonAction() {
-        self.dismiss(animated: true, completion: nil)
+        guard let panel = children.first as? UINavigationController else { return }
+
+        switch bookmarkPanelState {
+        case .home:
+            self.dismiss(animated: true, completion: nil)
+
```
Comment (nbhasin2) [inline Client/Frontend/Library/LibraryViewController/LibraryViewController.swift:316]:
```suggestion
```
nit: extra space

File: Client/Frontend/Library/LibraryViewController/LibraryViewController.swift [hunk 5]
```diff
@@ -217,22 +231,110 @@ class LibraryViewController: UIViewController {
...
+        updateBookmarkPanelState()
+        topLeftButtonSetup()
+        topRightButtonSetup()
+    }
+
+    func updateBookmarkPanelState() {
+        if librarySegmentControl.selectedSegmentIndex == 0 {
+            if let panel = children.first as? UINavigationController,
+               panel.viewControllers.count > 1 {
+                if bookmarkPanelState == .home {
```
Comment (nbhasin2) [inline Client/Frontend/Library/LibraryViewController/LibraryViewController.swift:244]: How do you feel about using guards or cleaning nested if lets? 🤔 
^maybe for later

File: Client/Frontend/Library/LibraryViewController/LibraryViewController.swift [hunk 6]
```diff
@@ -217,22 +231,110 @@ class LibraryViewController: UIViewController {
...
+            self.dismiss(animated: true, completion: nil)
+        case .inFolder:
+            guard let bookmarksPanel = panel.viewControllers.last as? BookmarksPanel else { return }
+            bookmarkPanelState = .inFolderEditMode
+            bookmarksPanel.enableEditMode()
+        case .inFolderEditMode:
+            guard let bookmarksPanel = panel.viewControllers.last as? BookmarksPanel else { return }
+            bookmarkPanelState = .inFolder
+            bookmarksPanel.disableEditMode()
+
```
Comment (nbhasin2) [inline Client/Frontend/Library/LibraryViewController/LibraryViewController.swift:324]:
```suggestion
```
nit : extra line

File: Client/Frontend/Library/LibraryViewController/LibraryViewController.swift [hunk 7]
```diff
@@ -217,22 +231,110 @@ class LibraryViewController: UIViewController {
...
+        case .bookmarkEditMode:
+            guard let bookmarkEditView = panel.viewControllers.last as? BookmarkDetailPanel else { return }
+            bookmarkEditView.save().uponQueue(.main) { _ in
+                self.bookmarkPanelState = .inFolderEditMode
+                panel.popViewController(animated: true)
+
```
Comment (nbhasin2) [inline Client/Frontend/Library/LibraryViewController/LibraryViewController.swift:330]:
```suggestion
```
nit extra line

**Review by nbhasin2 — APPROVED** [review]

File: Client/Frontend/Library/LibraryViewController/LibraryViewController.swift (hunk 5 above)
Comment (electricRGB) [reply Client/Frontend/Library/LibraryViewController/LibraryViewController.swift:244]: I had `guard let` there at one point. I think I just accidentally undid it. I can change that.

**Comment by electricRGB** [comment]
BR fails on `testDownloadFile` because http://www.ovh.net/files is inaccessible. Will merge as this is not an issue with our code.

**Pull request merged**